streams/src/main/java/org/apache/kafka/streams/processor/api/Processor.java

```diff
@@ -46,12 +47,12 @@
     default void init(final ProcessorContext<KOut, VOut> context) {}
 
     /**
-     * Process the record with the given key and value.
+     * Process the record. Note that record metadata is undefined in cases such as a forward call from a punctuator.
      *
-     * @param key the key for the record
-     * @param value the value for the record
+     * @param record the record to process
+     * @param recordMetadata the metadata of the record, if it is defined
      */
-    void process(KIn key, VIn value);
+    void process(Record<KIn, VIn> record, Optional<RecordMetadata> recordMetadata);
```

Comment:
Here's what the updated `Processor` API looks like. The "data" fields are in the `Record` (note that `ProcessorContext#forward` also accepts this class), and the "context" fields are in `RecordMetadata`, which is Optional, since it's undefined sometimes (like during a punctuation).